### ISSUE TITLE
Introduce non-libstd socket builder methods

### ISSUE THREAD

**Issue by kleimkuhler** [tokio-rs/mio#1211]
## Motivation

#1029 removed the `net` resources APIs to match those of `libstd`. This change also allowed Mio to remove it's dependency on `net2` which was used to provide these methods.

As part of the Mio update in Tokio--and mentioned briefly in the other issues--these methods would be helpful to provide again. For example, they are currently [removed](https://github.com/tokio-rs/tokio/pull/1767/files#diff-91404b27ce0ae7138a049039d5674df7L365-L510) as part of the Mio update PR in Tokio.

`socket2-rs`/`net2` provide a lot of additional methods that Mio does not really need and reintroduces a dependency when the syscalls would be fairly straightforward to implement directly. That being said, it's a dependency that can be relied on for the things we would need it for.

Alternatively, a new API could be introduced that provides builder methods for sockets (such as `set_keepalive`/`keepalive`/`set_ttl`/etc.) This approach would not introduce a new dependency. Something along the lines of providing a `TcpSocket` that has these methods and is consumed with a `connect -> TcpStream`.

Speaking with @carllerche offline about this I think the second approach of introducing something along the lines of `TcpSocket` builder is what Mio should move forward with, but there is more design conversation that should go into it.

**Comment by Thomasdezeeuw** [comment]
I've outed thing opinion before, but I'll repeat it here. I don't think it wise to duplicate the effort of a `net2`/`socket` crate. We already have the `FromRawFd`/`FromRawSocket` and `from_std` implementations that allow for advanced usage of all our `net` type (not just `TcpStream`, as the other types also have useful options that the std lib versions don't provide).

In the past I've also tried adding a way to set options on `TcpListener`, see https://github.com/tokio-rs/mio/pull/1068. I think it would be best if we would add the required methods to std lib instead, then everybody can benefit.
